corehq/apps/case_search/xpath_functions/query_functions.py

```diff
@@ -114,3 +114,17 @@ def match_all(node, context):
             serialize(node)
         )
     return filters.match_all()
+
+
+def match_none(node, context):
+    if len(node.args):
+        raise XPathFunctionException(
+            _("'match-none()' does not take any arguments"),
+            serialize(node)
+        )
+    return filters.match_none()
+
+
+def or_(node, context):
```

Comment:
You wrote earlier
> Currently an empty OR or OR with empty strings only throws an error

Now that we're going back to this, I just wanted to double check - is that a 500 error or a user-facing error?  Wondering if it's worth putting in a `if not node.args: raise XPathFunctionException`

Reply:
I was mostly thinking about whether it's an error that will be captured by this `except` statement:
https://github.com/dimagi/commcare-hq/blob/f51b264bfbe167b0440e67da464639bfa7e54131/corehq/apps/ota/views.py#L127-L130
and be logged as a user error, or 500 and get reported in sentry as a bug.  Would you mind adding a test like this one just to document the expected behavior?
https://github.com/dimagi/commcare-hq/blob/f51b264bfbe167b0440e67da464639bfa7e54131/corehq/apps/case_search/tests/test_case_search_filters.py#L180-L186

Reply:
Yeah, that's the case search endpoint - that's where it catches certain exceptions and returns a 400 response to formplayer.  Not sure what happens on the formplayer side or how that's displayed to the user, though.  There's also [this code](https://github.com/dimagi/commcare-hq/blob/f51b264bfbe167b0440e67da464639bfa7e54131/corehq/apps/case_search/utils.py#L73-L83) that catches `CaseFilterError` and re-raises it as `CaseSearchUserError` for the same reason, though I just noticed that it does log this type of error in sentry as well.  Looks like that's also caught [here](https://github.com/dimagi/commcare-hq/blob/f51b264bfbe167b0440e67da464639bfa7e54131/corehq/apps/reports/standard/cases/case_list_explorer.py#L40-L40) for the CLE.

Reply:
Oof, I hadn't seen that block, that exception handling seems overbroad.  `XPathFunctionException` _is_ a type of `CaseFilterError`, so anything that catches the latter will catch the former:
https://github.com/dimagi/commcare-hq/blob/f51b264bfbe167b0440e67da464639bfa7e54131/corehq/apps/case_search/exceptions.py#L24-L24

Reply:
> 400s are "something is wrong with the request" sort of error which is more suitable for this kind of issue

Well put.  400 means "this error is expected, the code is working, and this is the user's fault", while 500 means "the server is broken"